### PR TITLE
63 ip to hostname

### PR DESCRIPTION
**Pull request by heidiklum**
Refactored the code to use "host" instead of "ip" in the definition of a node.

For example, where the proto used to be "NodeAddress {uint32 id = 1; string ip = 2; uint32 port = 3;}", it is now  "NodeAddress {uint32 id = 1; string host = 2; uint32 port = 3;}".

Similarly, the arguments are now "--host" and --"knownHost" instead of "--ip" and "--targetIp", respectively.  Note that I also replaced "target*" with "known*", so "--knownHost", "--knownPort" and - if used optionally - "--knownId".

This obviously caused all the callers to have to be refactored so it cased changes in client.js and index.js.  All of these were fixed and tested, with the two drivers attached.

[driver_hash_manual.sh.txt](https://github.com/bushidocodes/chord-grpc/files/3855259/driver_hash_manual.sh.txt)

[driver_hash_npm_manual.sh.txt](https://github.com/bushidocodes/chord-grpc/files/3855261/driver_hash_npm_manual.sh.txt)

(Please note that GitHub rejects the ".sh" extension so just appended ".txt".  Then you'd have to "chmod u+x driver_*.sh" after renaming.)

In both drivers, just change the "WORKING_DIR" constant near the very top.  The "driver_hash_manual" calls "node node ..." directly, whereas the "driver_hash_npm_manual" calls "npm run devServer -- ..." (or "devClient").

I couldn't unfortunately get the "docker-compose up" call to work although I tried to modify the .yaml file.

In an attempt to fix the docker problem, I added a new feature to the main() of node.js to easily compute the hash of a value from the command line.  So, if you call "node node --hashOnly node0", it will return that the ID would be "2" - output is "ID { 2 } computed from hash of { node0 }".

Using that, I created a list of node names that wouldn't cause collisions. The list is pasted as a comment in the npm driver file:

"# 0 --> node9; 1 --> node2; 2 --> node0; 3 --> node6; 4 --> node4; 5 --> node3; 6 --> node7; 7 --> node1"

In other words, those results are that if "node9" is indeed the only thing hashed, it will create a node with ID 0 - in other words, that's the hash of "node9", not the hash of "node9:84xx".

Similarly, there is a list of deconflicted port numbers for convenience:

"# 0 --> 8441; 1 --> 8448; 2 --> 8454; 3 --> 8444; 4 --> 8446; 5 --> 8450; 6 --> 8451; 7 --> 8440"

Note, however, that these are intended to be hashed in conjuction with the host.  So, "localhost:8441" is what yields a node with ID 0.

This pull request should close issue #63 .

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -22,25 +22,25 @@ npm install
 Then run start the first node:
 
 ```sh
-npm run devServer -- --ip localhost --port 8440 --id 0 --targetIp localhost --targetPort 8440 --targetId 0
+npm run devServer -- --host localhost --port 8440 --knownHost localhost --knownPort 8440
```
Comment (bushidocodes) [inline README.md:25]: Much improved semantics! 👏

File: client.js [hunk 2]
```diff
@@ -224,10 +228,15 @@ function main() {
...
+        let crawler = new ChordCrawler(
+          target.host,
+          target.port,
+          CRAWLER_INTERVAL_MS
+        );
         const express = require("express");
         const app = express();
-        const port = args.webPort || 3000;
+        // TBD is this OR a bug?
+        const port = args.webPort || CRAWLER_INTERVAL_MS;
```
Comment (bushidocodes) [inline client.js]: This is a bug. They were just using the same magic number. This is the default port that the Express web service listens on.

File: node.js [hunk 3]
```diff
@@ -1444,9 +1444,32 @@ async function main() {
   // syntactic sugar for arguments
   const args = minimist(process.argv.slice(2));
 
+  // kludge to deconflict node IDs from hashed values
```
Comment (bushidocodes) [inline node.js:1447]: Just to clarify. You see this as a temporary stopgap until we increase the size of our hash table, right?

File: commands.md [hunk 4]
```diff
@@ -1,28 +1,28 @@
...
+npm run devServer -- --host localhost --port 8448 --knownHost localhost --knownPort 8440
+npm run devServer -- --host localhost --port 8450 --knownHost localhost --knownPort 8440
 
 <!-- Alvaro: npm run devClient worked for me, but it does not ends so I can't run the following command -->
 
-node client insert --ip localhost --port 8440 --id 2
-node client insert --ip localhost --port 8440 --id 5 --displayName "Alvaro is cool" --reputation 99 --aboutMe "I'm so cool I need no description"
-node client lookup --ip localhost --port 8440 --id 2
-node client remove --ip localhost --port 8440 --id 2
+node client insert --host localhost --port 8440
```
Comment (bushidocodes) [inline commands.md]: `id` has been a bit overloaded for us, but these values are actually the `id` of the user that we're inserting, updating, looking up, removing, etc. So these actions fail currently. You need to put the id arguments back for 14-17

**Comment by bushidocodes** [comment]
I used your hashing ID helper to plumb out the `docker-compose.yml`


```yml
version: "3"
services:
  node0:
    build:
      context: .
      dockerfile: NodeDockerfile
    # image: chordnode
    # container_name: chordnode
    command: npm run devServer -- --host node0 --knownHost node0 --knownPort 1337
  node2:
    build:
      context: .
      dockerfile: NodeDockerfile
    # image: chordnode
    # container_name: chordnode
    command: npm run devServer -- --host node2 --knownHost node0 --knownPort 1337
    depends_on:
      - node0
  node3:
    build:
      context: .
      dockerfile: NodeDockerfile
    # image: chordnode
    # container_name: chordnode
    command: npm run devServer -- --host node3 --knownHost node0 --knownPort 1337
    depends_on:
      - node0
  node4:
    build:
      context: .
      dockerfile: NodeDockerfile
    # image: chordnode
    # container_name: chordnode
    command: npm run devServer -- --host node4 --knownHost node0 --knownPort 1337
    depends_on:
      - node0
  node6:
    build:
      context: .
      dockerfile: NodeDockerfile
    # image: chordnode
    # container_name: chordnode
    command: npm run devServer -- --host node6 --knownHost node0 --knownPort 1337
    depends_on:
      - node0
  node9:
    build:
      context: .
      dockerfile: NodeDockerfile
    # image: chordnode
    # container_name: chordnode
    command: npm run devServer -- --host node9 --knownHost node0 --knownPort 1337
    depends_on:
      - node0
  node11:
    build:
      context: .
      dockerfile: NodeDockerfile
    # image: chordnode
    # container_name: chordnode
    command: npm run devServer -- --host node11 --knownHost node0 --knownPort 1337
    depends_on:
      - node0
  node15:
    build:
      context: .
      dockerfile: NodeDockerfile
    # image: chordnode
    # container_name: chordnode
    command: npm run devServer -- --host node15 --knownHost node0 --knownPort 1337
    depends_on:
      - node0
  client:
    build:
      context: .
      dockerfile: ClientDockerfile
    # image: chordclient
    # container_name: chordclient
    ports:
      - "1337:1337"
    depends_on:
      - node0
    # volumes:
    #   - .:/home/node/app
    #   - /home/node/app/node_modules/
    command: npm run devClient -- crawl --host node0 --port 1337 --webPort 1337
```

image

**Review by bushidocodes — CHANGES_REQUESTED** [review]
Good refactor. I have a few requested changes. Please also try to use the docker-compose I pasted into a comment. Thank you for this!

File: node.js (hunk 3 above)
Comment (heidiklum) [reply node.js:1447]: Yup.  Although I'm sure we'll want to keep it around long after because it is handy.  So, I tagged it as a "kludge" currently because I'm sure there is a more elegant way to achieve this  . . . and it will likely stay a kludge since we'll like using it for debug and never did the more elegant way :-)  Or, maybe this isn't bad in which case it shouldn't be marked as a kludge.

File: client.js (hunk 2 above)
Comment (heidiklum) [reply client.js]: Understood.  Replaced instead with "DEFAULT_HOST_PORT".

File: node.js (hunk 3 above)
Comment (heidiklum) [reply node.js:1447]: (in other words, i'll leave it like this in my next commit but am happy to revisit one way or the other without hurting my feelings)

File: commands.md (hunk 4 above)
Comment (heidiklum) [reply commands.md]: reverted as requested

Comment (heidiklum) [reply commands.md]: reverted as requested

**Comment by heidiklum** [comment]
responded to all comments and did a new commit

**Review by bushidocodes — APPROVED** [review]
Great work! 🙌 LGTM!
